harness/determined/pytorch/_pytorch_context.py

```diff
@@ -57,4 +103,344 @@ def get_lr_scheduler(self) -> Optional[pytorch.LRScheduler]:
+                be constructed, allowing to compute higher order derivative
+                products. Defaults to ``False``.
+        """
+        if self._use_amp:
+            check.eq(
+                loss,
+                self._loss,
+                "Only support configure apex amp on distributed training with only one loss",
+            )
+            self._loss = loss
+
+            with apex.amp.scale_loss(loss, self.optimizers) as scaled_loss:
+                scaled_loss.backward(
+                    gradient=gradient, retain_graph=retain_graph, create_graph=create_graph
+                )
+
+                if self.hvd_config.use and self._should_communicate_and_update():
+                    # When using horovod, we need to finish communicating gradient
+                    # updates before they are unscaled which happens when we exit
+                    # of this context manager.
+                    #
+                    # Unfortunately, there is no clean way to support unscaling
+                    # happening after synchronizing the optimizer on apex.
+                    # A short-term solution is to synchronize all of the optimizers
+                    # every time backward() is called, which slows down the training
+                    # quite a lot especially if the size of all the parameters
+                    # owned by all the optimizers is large. A long-term solution is
+                    # to integrate torch native AMP (https://pytorch.org/docs/stable/amp.html),
+                    # which will come out soon.
+                    for optimizer in self.optimizers:
+                        optimizer.synchronize()
+        else:
+            loss.backward(  # type: ignore
+                gradient=gradient, retain_graph=retain_graph, create_graph=create_graph,
+            )
+
+    def _average_gradients(self, parameters: Any, divisor: int) -> None:
+        check.gt_eq(divisor, 1)
+        if divisor == 1:
+            return
+
+        divisor_value = float(divisor)
+        for p in filter(lambda param: param.grad is not None, parameters):
+            p.grad.data.div_(divisor_value)
+
+    def step_optimizer(
+        self,
+        optimizer: torch.optim.Optimizer,  # type: ignore
+        clip_grads: Optional[Callable[[Union[Iterable[torch.Tensor], torch.Tensor]], None]] = None,
+    ) -> None:
+        """
+        Perform a single optimization step.
+
+        This function must be called once for each optimizer. However, the order of
+        different optimizers' optimization steps can be specified by calling
+        this function in different orders. Also, gradient accumulation across iterations
+        is performed by the Determined training loop by setting the experiment configuration
+        optimization.aggregation_frequency.
+
+        Arguments:
+            optimizer(``torch.optim.Optimizer``): An optimization step of which optimizer should
+                be performed.
```

Comment:
"An optimization step of which optimizer should be performed" -> "which optimizer should be stepped"